File: PHP/crypto/aes/Encrypt.php
<?php

class Encrypt {

    private $key;
    private $iv;
    private $clearText;
    private $encryptedText;

    /**
     * Class constructor
     * Generates the key, initialization vector, and encrypts the data
     * @param string $clearText
     */
    public function __construct($clearText) {
        $this->clearText = $clearText;
        $this->key = $this->generateKey();
        $this->iv = $this->generateIv();
        $this->encryptedText = $this->encrypt();
    }

    /**
     * Generates the key used for encryption
     * @return string Secure random bytes as a string
     */
    private function generateKey() {
        return openssl_random_pseudo_bytes(32);
    }

    /**
     * Generates the initialization vector to be used
     * @return string Initialization vector
     */
    private function generateIv() {
        $size = mcrypt_get_iv_size(MCRYPT_RIJNDAEL_128, MCRYPT_MODE_CBC);
        return mcrypt_create_iv($size, MCRYPT_DEV_URANDOM);
    }
    
    /**
     * Encrypts the data
     * @return string Encrypted data
     */
    private function encrypt() {
        return mcrypt_encrypt(MCRYPT_RIJNDAEL_128, $this->key, $this->clearText, MCRYPT_MODE_CBC, $this->iv);
    }

    /**
     * Gets the key that was used to encrypt the data
     * @return string Key used for encryption/decryption
     */
    public function getKey() {
        return $this->key;
    }

    /**
     * Gets the initialization vector that was used when the data was encrypted
     * @return string Initialization vector used during the encryption process
     */
    public function getIv() {
        return $this->iv;
    }

    /**
     * Returns the encrypted data as a base64 string
     * @return string The encrypted data in base64 format
     */
    public function getEncoded() {
        return base64_encode($this->encryptedText);
    }
    
    /**
     * Returns the encrypted data in binary format
     * @return string The encrypted data
     */
    public function getBinary() {
        return $this->encryptedText;
    }
    
    /**
     * Overrides the native method to return the encrypted data as a printable string
     * @return string The encrypted data as a printable string
     */
    public function __toString() {
        return $this->getEncoded();
    }

}

?>

File: Java/src/com/aaronjwood/multilib/crypto/aes/Encrypt.java
package com.aaronjwood.multilib.crypto.aes;

import javax.crypto.*;
import javax.crypto.spec.IvParameterSpec;
import java.io.UnsupportedEncodingException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Base64;

public final class Encrypt {

    private Cipher encryptCipher;
    private SecretKey key;
    private IvParameterSpec iv;

    private byte[] clearText;
    private byte[] encryptedText;

    /**
     * Encrypts data using AES-128
     *
     * @param clearText The data to be encrypted
     * @throws java.security.NoSuchAlgorithmException
     * @throws javax.crypto.NoSuchPaddingException
     * @throws java.security.InvalidKeyException
     * @throws javax.crypto.BadPaddingException
     * @throws java.security.InvalidAlgorithmParameterException
     * @throws javax.crypto.IllegalBlockSizeException
     * @throws java.io.UnsupportedEncodingException
     */
    public Encrypt(String clearText) throws NoSuchAlgorithmException, NoSuchPaddingException, InvalidKeyException, InvalidAlgorithmParameterException, UnsupportedEncodingException, IllegalBlockSizeException, BadPaddingException {

        //Generate the IV and key
        this.iv = new IvParameterSpec(this.generateIv());
        this.key = this.generateKey();
        this.encryptCipher = createCipher();
        this.clearText = this.convertClearText(clearText);
        this.encryptedText = this.encrypt();
    }

    /**
     * Converts the clear text passed by the user to an array of bytes
     *
     * @param clearText The clear text passed by the user
     * @return The byte representation of the clear text
     */
    private byte[] convertClearText(String clearText) throws UnsupportedEncodingException {

        //Convert the clear text passed by the user into bytes
        return clearText.getBytes("UTF-8");
    }

    /**
     * Creates an AES cipher using CBC mode with PKCS5 padding
     *
     * @return The cipher used to encrypt data
     */
    private Cipher createCipher() throws NoSuchAlgorithmException, NoSuchPaddingException, InvalidKeyException, InvalidAlgorithmParameterException {

        //Create an AES cipher in CBC mode using PKCS5 padding
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.ENCRYPT_MODE, this.key, this.iv);
        return cipher;
    }

    /**
     * Generates a random IV to be used in the encryption process
     *
     * @return The IV's byte representation
     */
    private byte[] generateIv() {
        SecureRandom random = new SecureRandom();
        byte[] ivBytes = new byte[16];
        random.nextBytes(ivBytes);
        return ivBytes;
    }

    /**
     * Generates a secret key to be used in the encryption process
     *
     * @return The secret key
     */
    private SecretKey generateKey() throws NoSuchAlgorithmException {
        KeyGenerator keygen;

        //Java normally doesn't support 256-bit key sizes without an extra installation so stick with a 128-bit key
        keygen = KeyGenerator.getInstance("AES");
        keygen.init(128);
        SecretKey aesKey = keygen.generateKey();
        return aesKey;

    }

    /**
     * Returns the initialization vector
     *
     * @return The randomly generated IV
     */
    public IvParameterSpec getIv() {
        return this.iv;
    }

    /**
     * Returns the key used for encryption
     *
     * @return The randomly generated secret key
     */
    public SecretKey getKey() {
        return this.key;
    }

    /**
     * Encrypts the data passed during instantiation of this object
     *
     * @return The byte representation of the encrypted data
     */
    private byte[] encrypt() throws IllegalBlockSizeException, BadPaddingException {
        return this.encryptCipher.doFinal(this.clearText);

    }

    /**
     * Returns the encrypted text as a base64 encoded string
     *
     * @return The encrypted base64 encoded string
     */
    @Override
    public String toString() {
        return Base64.getEncoder().encodeToString(encryptedText);
    }

}
